google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,469 @@
+   * any other pending operations (typically any other `Read()` calls) complete
+   * and then call `Finish()` to find the status of the streaming RPC.
+   */
+  virtual future<bool> Write(Request const&, grpc::WriteOptions) = 0;
+
+  /**
+   * Return the final status of the streaming RPC.
+   *
+   * The application must wait until all pending `Read()` and `Write()`
+   * operations have completed before calling `Finish()`.
+   */
+  virtual future<Status> Finish() = 0;
+};
+
+template <typename RequestType, typename ResponseType>
+class ResumableAsyncStreamingReadWriteRpcImpl
+    : public ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType> {
+ public:
+  ResumableAsyncStreamingReadWriteRpcImpl(
+      std::shared_ptr<RetryPolicy const> retry_policy,
+      std::shared_ptr<BackoffPolicy const> backoff_policy,
+      AsyncSleeper const sleeper,
+      AsyncStreamFactory<RequestType, ResponseType> stream_factory,
+      StreamInitializer<RequestType, ResponseType> initializer)
+      : retry_policy_prototype_(std::move(retry_policy)),
+        backoff_policy_prototype_(std::move(backoff_policy)),
+        sleeper_(std::move(sleeper)),
+        stream_factory_(std::move(stream_factory)),
+        initializer_(std::move(initializer)) {}
+
+  ResumableAsyncStreamingReadWriteRpcImpl(ResumableStreamingReadRpc&&) = delete;
+  ResumableAsyncStreamingReadWriteRpcImpl& operator=(
+      ResumableStreamingReadRpc&&) = delete;
+
+  future<Status> Start() override {
+    future<Status> status_future;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      status_future = status_promise_.get_future();
+    }
+    Initialize(retry_policy_prototype_->clone(),
+               backoff_policy_prototype_->clone());
+    return status_future;
+  }
+
+  future<absl::optional<ResponseType>> Read() override {
+    future<absl::optional<ResponseType>> read_future;
+    bool retrying = false;
+
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      retrying = stream_state_ == State::kRetrying;
+      if (retrying) {
+        read_reinit_done_.emplace(absl::make_optional(promise<void>()));
+        read_future = read_reinit_done_->get_future();
+      } else {
+        read_future = stream_->Read();
+        outstanding_read_ = true;
+      }
+    }
+
+    if (retrying) {
+      return read_future;
+    }
+
+    return read_future.then(
+        [this](future<absl::optional<Response>> optional_response_future) {
+          auto optional_response = optional_response_future.get();
+          bool response_good = optional_response.has_value();
+          bool should_finish = false;
+
+          {
+            std::lock_guard<std::mutex> g{mu_};
+            outstanding_read_ = false;
+
+            // This means that an unsuccessful `Write` occurred, but could not
+            // enter the retry loop because of this present outstanding `Read`.
+            if (stream_state_ == State::kMustRetry) {
```

Comment:
I don't think this state is needed. Instead, just move to retrying and chain the pending futures using then() to wait on them if they exist.

Reply:
isn't the pending future already returned to the user? would having `then` being called on the pending future in two different places require a shared future? or is this possible with `google::cloud::future`?